### DISCUSSION TITLE
How to remove discrepencies in a 2D pattern produced from the integration of an image calculated with `calcfrom2d` ?

### DISCUSSION THREAD

**Discussion by loichuder** [silx-kit/pyFAI#2258] [Q&A] [answered]
I want to write a test that checks that the integration 2D works as intended. For this, I am generating arrays for radial, azimuthal and the resulting integrated patterns:
```python
tth = numpy.linspace(0, 30, 200)
chi = numpy.linspace(-180, 180, 180)
intensity = numpy.full((chi.size, tth.size), 10)
```

I am then using `ai.calcfrom2d` to create an image whose 2D integration should give the `intensity` array:

```python
# ai is the azimuthal integrator
calc_image = ai.calcfrom2d(intensity, tth, chi, mask=ai.mask)
```

But the problem is that when I do the integration on this `calc_image`, I do get "holes" in the integrated pattern:
```
result = ai.integrate2d(calc_image, tth.size, chi.size)
integrated_intensity = result.intensity
```
![Figure_1](https://github.com/user-attachments/assets/08f88675-d19f-4711-95df-c0f38e9b3482)

I attached an HDF5 file that contain all the data for easier browsing: [test_pyFAI_2d.zip](https://github.com/user-attachments/files/16764199/test_pyFAI_2d.zip)

I guess some bins are empty but I do not really understand why ? How could I avoid this ? Should I create a more sparse `intensity` array ?

<details>
<summary>
Full Python script
</summary>
<pre>
import pyFAI
import pyFAI.detectors
from pyFAI.azimuthalIntegrator import AzimuthalIntegrator
import numpy
import matplotlib.pyplot as plt

geometry = {
    "dist": 10e-2,  # 10 cm
    "poni1": 10e-2,  # 10 cm
    "poni2": 10e-2,  # 10 cm
    "rot1": numpy.radians(10),  # 10 deg
    "rot2": 0,  # 0 deg
    "rot3": 0,  # 0 deg
}
ai = AzimuthalIntegrator(
    detector=pyFAI.detectors.detector_factory("Pilatus1M"), wavelength=1e-10, **geometry
)
tth = numpy.linspace(0, 30, 200)
chi = numpy.linspace(-180, 180, 180)
intensity = numpy.full((chi.size, tth.size), 10)

calc_image = ai.calcfrom2d(intensity, tth, chi, mask=ai.mask)

result = ai.integrate2d(calc_image, tth.size, chi.size)
integrated_intensity = result.intensity

f, ax = plt.subplots(ncols=2)
ax[0].pcolormesh(intensity)
ax[0].set_title("Original")
ax[1].pcolormesh(integrated_intensity)
ax[1].set_title("Integrated")

plt.show()

</pre>
</details>

**Comment by kif** [comment, marked as answer]
The holes you see are consequences of the gaps in the detector "Pilatus 1M" which are hard-coded in the `pyFAI.detectors` class.
https://github.com/silx-kit/pyFAI/blob/main/src/pyFAI/detectors/_dectris.py#L69
You can always re-define the detector's mask like:
```
detector = pyFAI.detector_factory("Pilatus1M")
detector.mask = numpy.zeros(detector.shape, "int8")
```

Note that the `calcfrom1/2d` does interpolation while integrate performs histogramming, the difference is subtle but you won't get back what you input initially.

**Reply by loichuder** [reply]
> The holes you see are consequences of the gaps in the detector "Pilatus 1M" 

Of course, that makes sense. I wonder why I didn't think about that :facepalm: 

Thanks!
